pkg/kapp/diff/change_factory.go

```diff
@@ -4,26 +4,76 @@
 package diff
 
 import (
+	"context"
+	"fmt"
 	ctlres "github.com/k14s/kapp/pkg/kapp/resources"
+	"k8s.io/apimachinery/pkg/types"
 )
 
 type ChangeFactory struct {
 	rebaseMods                               []ctlres.ResourceModWithMultiple
 	diffAgainstLastAppliedFieldExclusionMods []ctlres.FieldRemoveMod
+	resources                                ctlres.Resources
 }
 
+type ChangeFactoryFunc func(ctx context.Context, existingRes, newRes ctlres.Resource) (Change, error)
+
+var _ ChangeFactoryFunc = ChangeFactory{}.NewChangeSSA
+var _ ChangeFactoryFunc = ChangeFactory{}.NewChangeAgainstLastApplied
+var _ ChangeFactoryFunc = ChangeFactory{}.NewExactChange
+
 func NewChangeFactory(rebaseMods []ctlres.ResourceModWithMultiple,
-	diffAgainstLastAppliedFieldExclusionMods []ctlres.FieldRemoveMod) ChangeFactory {
+	diffAgainstLastAppliedFieldExclusionMods []ctlres.FieldRemoveMod, resources ctlres.Resources) ChangeFactory {
+
+	return ChangeFactory{rebaseMods, diffAgainstLastAppliedFieldExclusionMods, resources}
+}
+
+func (f ChangeFactory) NewChangeSSA(ctx context.Context, existingRes, newRes ctlres.Resource) (Change, error) {
+	if existingRes != nil {
+		historylessExistingRes, err := f.NewResourceWithHistory(existingRes).HistorylessResource()
+		if err != nil {
+			return nil, err
+		}
+
+		existingRes = historylessExistingRes
+	}
+
+	newResAsIs := newRes
+	if newResAsIs != nil {
+		historylessNewRes, err := f.NewResourceWithHistory(newRes).HistorylessResource()
+		if err != nil {
+			return nil, err
+		}
+
+		newResAsIs = historylessNewRes
+	}
 
-	return ChangeFactory{rebaseMods, diffAgainstLastAppliedFieldExclusionMods}
+	dryRunRes := newResAsIs
+	if dryRunRes != nil && existingRes != nil {
+		newResBytes, _ := newRes.AsYAMLBytes()
+		dryRunResult, err := f.resources.Patch(existingRes, types.ApplyPatchType, newResBytes, true)
+		if err != nil {
+			return nil, fmt.Errorf("SSA dry run: %s", err)
+		}
+		dryRunRes = dryRunResult
+	}
+
+	return NewChangeSSA(existingRes, newResAsIs, dryRunRes), nil
 }
 
-func (f ChangeFactory) NewChangeAgainstLastApplied(existingRes, newRes ctlres.Resource) (Change, error) {
+func (f ChangeFactory) NewChangeAgainstLastApplied(ctx context.Context, existingRes, newRes ctlres.Resource) (Change, error) {
 	// Retain original copy of existing resource and use it
 	// for rebasing last applied resource and new resource.
 	existingResForRebasing := existingRes
+	var err error
 
 	if existingRes != nil {
+		// Strip rebasing "base" object of kapp history so that it never shows up in the diff, regardless
+		// of rebase rules used
+		existingResForRebasing, err = f.NewResourceWithHistory(existingRes).HistorylessResource()
```

Comment:
I believe there was a sleeper bug here, which unfolded like following:
- By the end of this function `existingRes` was unconditionally stripped of history with `HistorylessResource()`
- `newRes` was unconditionally stripped of history with `HistorylessResource()`
- `existingResForRebasing` has history kept, because it comes from the cluster as is
- When building `rebasedNewRes` using `existingResForRebasing + newRes`, history was removed due to default rebase rules included in the kapp transplanting annotations from `newRes` 
- If we redeploy same object as previous deployment, end result has no spurious diff, because both objects have history stripped

Should provided config had different rebase rules, `rebasedNewRes` is going to have history annotations from `existingResForRebasing`, but `existingRes` wont, generating spurious diff.

Similar situation arises when `metadata.managedFields` is populated by the cluster. This field is considered history and is stripped in `HistorylessResource()` call, but default rebase rules have no mentions of it.

Instead of updating default rebase rule, I opted for fixing it at the source. Here I am stripping history from all objects, ensuring it is not going to show up in any diffs regardless of kapp config provided. Technically it is a change in behaviour, hence this comment to highlight it.

Reply:
to double check -- changes in NewChangeAgainstLastApplied are not really affecting SSA, right? (i might have to ask to separate them out into a diff PR).